### PR TITLE
Feature/217 project icon

### PR DESCRIPTION
**Pull request by waltersajtos**
## Description

A few sentences describing the overall goals of the pull request's commits.

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [x] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

Outline the steps to test or reproduce the PR here.
These steps will be used during release testing.

1. Add a new project
2. Drag and Drop/ browse an image in the box
3. An image should be previewed underneath the drop area and the file should start uploading
4. Create the project and check if the icon has been added to the project (in the database, it won't be displayed yet).

## Link to issue
[#266](https://github.com/DigitalExcellence/dex-frontend/issues/266)
Closes: #266

### CODE REVIEW

**Comment by MeesvanStraten** [comment]
@waltersajtos 
after i added the project and click on the project again, should the icon be visible? 
While adding a new project i am able to see the icon but not after confirming the project. There only is the placeholder.

**Review by MeesvanStraten — COMMENTED** [review]
@waltersajtos 
after i added the project and click on the project again, should the icon be visible?
While adding a new project i am able to see the icon but not after confirming the project. There only is the placeholder.

**Comment by waltersajtos** [comment]
> @waltersajtos
> after i added the project and click on the project again, should the icon be visible?
> While adding a new project i am able to see the icon but not after confirming the project. There only is the placeholder.

Yes, that is correct. It is a draft, the backend is not done yet so I can't fully implement it yet.

**Review by MeesvanStraten — DISMISSED** [review]
Looks good! and works as the issue describes.

**Review by MeesvanStraten — APPROVED** [review]

**Comment by Brend-Smits** [comment]
Please resolve the conflicts. Also, is this ready for review? I'm not sure how far the implementation is in terms of uploading the image.

**Comment by waltersajtos** [comment]
I'm currently working on changing the logic as requested, I'm changing when the images get uploaded (on save project button instead of instantly). I will implement the comments above when I got the logic working.

**Comment by waltersajtos** [comment]
> Found a (small) issue: when you add an image and remove it. It is not possible to add the same picture again.
This issue should be fixed with [my last commit](https://github.com/DigitalExcellence/dex-frontend/pull/292/commits/14c5d8ceb32b2ea7be41598d23a693b29e6e384b)

> It also would be nice to add some error messaging for non accepted file types.
This should be fixed with [this commit](https://github.com/DigitalExcellence/dex-frontend/pull/292/commits/6e3274516a62ed332c2fe584051796c34cb0e157)

**Comment by waltersajtos** [comment]
> Please resolve the conflicts. Also, is this ready for review? I'm not sure how far the implementation is in terms of uploading the image.

@Brend-Smits, the uploading works. I just need to do some refactoring/ cleaning up. It will be ready for review tomorrow.

**Comment by niraymak** [comment]
Looks good! I think it would be nice to also implement this on the update project page?

**Comment by waltersajtos** [comment]
> Looks good! I think it would be nice to also implement this on the update project page?

I agree, working on this now!

**Comment by waltersajtos** [comment]
The functionality is done. The design probably needs some work and the preview on the edit page still needs to be implemented, I couldn't do that because there is no way to load the project icons for now.

**Comment by niraymak** [comment]
For me nothing happens when I click on the Edit Project button. (saving changes after editing project)

**Comment by waltersajtos** [comment]
> For me nothing happens when I click on the Edit Project button. (saving changes after editing project)

Hmm, I thought it worked before I pushed.. I think it is because of the templates in the edit page

File: src/app/modules/project/edit/edit.component.ts [hunk 1]
```diff
@@ -130,23 +138,40 @@ export class EditComponent implements OnInit {
...
-        finalize(() => {
-          this.submitEnabled = false;
-        })
-      )
-      .subscribe(() => {
+    this.fileUploader.uploadFiles().subscribe(uploadedFiles => {
+      if (uploadedFiles[0]) {
+        edittedProject.fileId = uploadedFiles[0].id;
+        this.editProject(edittedProject);
+      } else {
```
Comment (niraymak) [inline src/app/modules/project/edit/edit.component.ts:145]: You should check if there is a file before uploading it. 

editProject should be called after uploading the file (if that is the case).

**Review by niraymak — CHANGES_REQUESTED** [review]

File: src/app/modules/project/edit/edit.component.ts (hunk 1 above)
Comment (waltersajtos) [reply src/app/modules/project/edit/edit.component.ts:145]: That should be checked indeed, I don't think that that's the problem tho.
fileUploader appears to be undefined because it's not rendered instantly because everything is in `<ng-container>` & `<ng-temaplate>`

Comment (waltersajtos) [reply src/app/modules/project/edit/edit.component.ts:145]: if uploadFiles() does not return anything if the file wasn't uploaded if that's what you mean

**Review by niraymak — DISMISSED** [review]

File: src/app/components/file-uploader/file-uploader.component.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      return of(undefined);
+    }
+  }
+
+  public setFiles(editFiles: Array<UploadFile>) {
+    editFiles.forEach(editFile => {
+      // Preview has to be changed when the infrastructure for showing the icons is in place.
+      this.files.push({
+        ...editFile,
+        preview: 'https://www.laurenillumination.com/wp-content/uploads/woocommerce-placeholder.png'});
```
Comment (Brend-Smits) [inline src/app/components/file-uploader/file-uploader.component.ts:137]: We should probably reference a png here from unDraw

File: src/app/modules/project/add/manual/manual.component.html [hunk 3]
```diff
@@ -66,6 +66,15 @@ <h3>Description*</h3>
           </div>
         </div>
       </div>
+      <div class="row">
+        <div class="col-12">
+          <div class="project-icon-picker">
+            <h3>Project icon</h3>
+            <app-file-uploader [acceptMultiple]="acceptMultiple" [acceptedTypes]="acceptedTypes">
```
Comment (Brend-Smits) [inline src/app/modules/project/add/manual/manual.component.html:73]: Why do we need the option to upload multiple file uploads for a project icon?

File: src/app/modules/project/edit/edit.component.html [hunk 4]
```diff
@@ -21,11 +21,11 @@
   <div class="offset-2 col-8">
     <h1>Edit project</h1>
 
-    <ng-container *ngIf="projectLoading; else projectLoaded">
+    <ng-container *ngIf="projectLoading">
```
Comment (Brend-Smits) [inline src/app/modules/project/edit/edit.component.html:24]: What's the reason for this change?

**Review by Brend-Smits — COMMENTED** [review]
I've asked @rubenb994 if he has time to take a look at this as well. He's more experienced with Angular side of things. Functionally I have yet to test this. Hopefully I will be able to do so soon. 

Sorry for the delay, schedule has been crazy :(

File: src/app/modules/project/add/manual/manual.component.html (hunk 3 above)
Comment (waltersajtos) [reply src/app/modules/project/add/manual/manual.component.html:73]: In the backend discussion, Max & Ruben mentioned that we probably want to support upload .pdf, .doc files, and such later on. I figured I may as well implement it now.

File: src/app/modules/project/edit/edit.component.html (hunk 4 above)
Comment (waltersajtos) [reply src/app/modules/project/edit/edit.component.html:24]: I had to switch out all the templates from that page because else the FileComponent wouldn't bind in any of the lifecycle hooks. If anyone has a better approach to this I would be happy to change is, I don't like the way it is now either but it was the only way I got it to work.

**Comment by rubenb994** [comment]
Okay finally got some time to look at this PR. First off the input is not styled and aligned properly. 
image
The text in the button is not centered, the two elements are not centered with in the container. The preview of the image is definitely not round. The font for the file name is quite large.  Why are the file size and the file size indicator (kb) on different lines? 

My suggestion is have the 'Drag drop files here' and the button side by side horizontally instead of vertically.

File: src/app/components/file-uploader/file-uploader.component.html [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        <div class="file-info">
+            <img alt="image-preview" class="file-icon" [src]=file?.preview >
+            <h4 class="file-name">
+                {{ file?.name }}
+            </h4>
+            <h4 class="file-size">
+                {{ file?.readableSize }}
+            </h4>
+
+            <div class="delete" (click)="deleteFile(i)">
```
Comment (rubenb994) [inline src/app/components/file-uploader/file-uploader.component.html:19]: Take a look at how the remove collaborators is styled, I think it would be nice if the remove file button would be the same

File: src/app/components/file-uploader/file-uploader.component.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
+import { Component, Input, ViewChild, ElementRef, OnInit } from '@angular/core';
+import { UploadFile } from 'src/app/models/domain/uploadFile';
+import { FileUploaderService } from 'src/app/services/file-uploader.service';
+import { map } from 'rxjs/operators';
+import { HttpEventType } from '@angular/common/http';
+import { forkJoin, Observable, of } from 'rxjs';
+import { AlertConfig } from '../../models/internal/alert-config';
```
Comment (rubenb994) [inline src/app/components/file-uploader/file-uploader.component.ts:7]: Please use imports with only absolute paths like `import { UploadFile } from 'src/app/models/domain/uploadFile';`

File: src/app/components/file-uploader/file-uploader.component.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+import { AlertType } from '../../models/internal/alert-type';
+import { AlertService } from '../../services/alert.service';
+
+@Component({
```
Comment (rubenb994) [inline src/app/components/file-uploader/file-uploader.component.ts:11]: Component documentation is missing, what is the functionality/purpose of this component?

File: src/app/components/file-uploader/file-uploader.component.ts [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  templateUrl: './file-uploader.component.html',
+  styleUrls: ['./file-uploader.component.scss']
+})
+export class FileUploaderComponent {
+
+  @Input() acceptMultiple: boolean;
+  @Input() acceptedTypes: Array<String>;
+  @ViewChild('fileDropRef') fileInput: ElementRef;
+
+  maxFileSize = 5242880;
```
Comment (rubenb994) [inline src/app/components/file-uploader/file-uploader.component.ts:22]: No all properties/methods have their proper access modifiers. The order of the properties/methods should follow the guidelines defined here: https://github.com/DigitalExcellence/dex-frontend/wiki/Coding-Guidelines#-methods-and-variables-order

File: src/app/components/file-uploader/file-uploader.component.ts [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+      return of(this.files);
+    } else {
+      // If there are no files in the list return undefined
+      return of(undefined);
+    }
+  }
+
+  public setFiles(editFiles: Array<UploadFile>) {
+    editFiles.forEach(editFile => {
+      // TODO: Preview has to be changed when the infrastructure for showing the icons is in place.
```
Comment (rubenb994) [inline src/app/components/file-uploader/file-uploader.component.ts:151]: Reference the issue which will solve this todo in the code

File: src/app/components/file-uploader/file-uploader.component.ts [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+          this.alertService.pushAlert(alertConfig);
+          this.deleteFile(this.files.indexOf(file));
+        }
+      } else {
+        const alertConfig: AlertConfig = {
+          type: AlertType.danger,
+          preMessage: 'This file is too big',
+          mainMessage: `File ${file.name} is too big, the max size is 5mb`,
+          dismissible: true,
+          timeout: this.alertService.defaultTimeout
```
Comment (rubenb994) [inline src/app/components/file-uploader/file-uploader.component.ts:82]: You can leave the timeout property unset if you are using the default timeout. It is a nullable property and the service will already use the default timeout if none is set.

File: src/app/models/domain/project.ts [hunk 11]
```diff
@@ -28,4 +29,5 @@ export interface Project {
   updated: Date;
   uri?: string;
   collaborators?: Collaborator[];
+  projectIcon: UploadFile;
```
Comment (rubenb994) [inline src/app/models/domain/project.ts:32]: This should be a nullable property

File: src/app/models/resources/project-add.ts [hunk 12]
```diff
@@ -23,4 +23,5 @@ export interface ProjectAdd {
   shortDescription: string;
   description?: string;
   url: string;
+  fileId: number;
```
Comment (rubenb994) [inline src/app/models/resources/project-add.ts:26]: This should be a nullable property

File: src/app/models/resources/project-update.ts [hunk 13]
```diff
@@ -24,5 +24,6 @@ export interface ProjectUpdate {
   name: string;
   shortDescription: string;
   description: string;
+  fileId: number;
```
Comment (rubenb994) [inline src/app/models/resources/project-update.ts:27]: This should be a nullable property

File: src/app/modules/project/add/manual/manual.component.html (hunk 3 above)
Comment (rubenb994) [reply src/app/modules/project/add/manual/manual.component.html:73]: Indeed it is really nice to have functionality. Even though we are not using it at the moment, it ensures in the future that we can support multi file uploads in a different part of the app.

File: src/app/modules/project/add/manual/manual.component.ts [hunk 14]
```diff
@@ -61,6 +63,13 @@ export class ManualComponent implements OnInit {
    */
   public modulesConfigration = QuillUtils.getDefaultModulesConfiguration();
 
+  /**
+   * Configuration for file-picker
+   */
+  public acceptedTypes: Array<string> = [ 'image/png', 'image/jpg', 'image/jpeg' ];
+  public acceptMultiple: Boolean = false;
```
Comment (rubenb994) [inline src/app/modules/project/add/manual/manual.component.ts:70]: You can omit the types when they can be automatically determined from the initialization. See: https://github.com/DigitalExcellence/dex-frontend/wiki/Coding-Guidelines#-type-definition

File: src/app/modules/project/add/manual/manual.component.ts [hunk 15]
```diff
@@ -123,20 +132,41 @@ export class ManualComponent implements OnInit {
...
-        this.router.navigate([`/project/overview`]);
-      });
+        .post(newProject)
+        .pipe(finalize(() => (this.submitEnabled = false)))
+        .subscribe(() => {
+          const alertConfig: AlertConfig = {
+            type: AlertType.success,
+            mainMessage: 'Project was succesfully saved',
+            dismissible: true,
+            timeout: this.alertService.defaultTimeout
```
Comment (rubenb994) [inline src/app/modules/project/add/manual/manual.component.ts:165]: You can remove the timeout property from the object when using the default value

File: src/app/modules/project/edit/edit.component.html [hunk 16]
```diff
@@ -132,15 +141,15 @@ <h3>Collaborators</h3>
...
         </div>
-      </ng-container>
-      <ng-template #noProject>
+      </div>
+      <div [hidden]="project">
         <p>No project could be found with id: {{invalidId}}</p>
-      </ng-template>
-    </ng-template>
+      </div>
+    </div>
```
Comment (rubenb994) [inline src/app/modules/project/edit/edit.component.html:150]: Please remove these empty lines

File: src/app/modules/project/edit/edit.component.html [hunk 17]
```diff
@@ -21,11 +21,11 @@
   <div class="offset-2 col-8">
     <h1>Edit project</h1>
 
-    <ng-container *ngIf="projectLoading; else projectLoaded">
+    <ng-container *ngIf="projectLoading">
       <p>Project is being loaded</p>
     </ng-container>
-    <ng-template #projectLoaded>
-      <ng-container *ngIf="project; else noProject">
+    <div [hidden]="projectLoading">
```
Comment (rubenb994) [inline src/app/modules/project/edit/edit.component.html:27]: This syntax has my preference of using here instead of the hidden:
```ts
    <ng-container *ngIf="someBoolean; else elseTemplate">
      
    </ng-container>
    <ng-template #elseTemplate>
      
    </ng-template>
```
The reason being; a `NG if else` ensures the items are not even rendered in the DOM, while hidden still creates the elements in the DOM but just hides them.

File: src/app/modules/project/edit/edit.component.ts [hunk 18]
```diff
@@ -130,25 +139,45 @@ export class EditComponent implements OnInit {
...
+            finalize(() => {
+              this.submitEnabled = false;
+            })
+        )
+        .subscribe(() => {
+          const alertConfig: AlertConfig = {
+            type: AlertType.success,
+            mainMessage: 'Project was succesfully updated',
+            dismissible: true,
+            timeout: this.alertService.defaultTimeout
```
Comment (rubenb994) [inline src/app/modules/project/edit/edit.component.ts:174]: Same timeout remark as before

File: src/app/components/file-uploader/drag-and-drop.directive.ts [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+  Input,
+  EventEmitter,
+  HostBinding,
+  HostListener
+} from '@angular/core';
+
+@Directive({
+  selector: '[app-drag-and-drop]'
+})
+export class DndDirective {
```
Comment (rubenb994) [inline src/app/components/file-uploader/drag-and-drop.directive.ts:13]: Please rename this to DragAndDropDirective. The name 'DndDirective' might be easy to remember now, but in the long run when the project gets bigger and bigger, this name will become more and more confusing.

**Review by rubenb994 — CHANGES_REQUESTED** [review]
Cool implementation, especially with the drag and dropping of files. But there are styling errors and I left some remarks on your code

File: src/app/modules/project/edit/edit.component.html (hunk 17 above)
Comment (waltersajtos) [reply src/app/modules/project/edit/edit.component.html:27]: It was like that before. I changed it because I could not find a way to access the viewChild if it's in an ng-template or ng-container because it's not accessible in ngOnInit. If you have any workaround for this I would love to hear it!

**Comment by waltersajtos** [comment]
I think I corrected everything, I left the edit page as it is (see comment on your [original message](https://github.com/DigitalExcellence/dex-frontend/pull/292#discussion_r499555403))

**Review by rubenb994 — DISMISSED** [review]

**Review by MeesvanStraten — DISMISSED** [review]
Tested and works great :) nice job

**Review by Brend-Smits — CHANGES_REQUESTED** [review]
Wait for me to merge this one please.

**Review by Ruby77 — COMMENTED** [review]
First merge this branch on this branch https://github.com/DigitalExcellence/dex-frontend/pull/331

**Comment by waltersajtos** [comment]
This should be ready for final review now.

**Review by Brend-Smits — CHANGES_REQUESTED** [review]
Functionally this works like a charm, good work. 
However, I think we should look into improving the visuals. Take this example:
 - When you upload an icon, why should the background color still be set to orange? I think the background color should be unset when an icon is uploaded. On all pages, project detail - overview but also top highlight cards. 
 - Icons should probably be stretched to fill the full circle that is provided by us so we don't get weird-looking UI if someone does not upload a proper size icon. 

I think we should pick these comments up in a separate issue, can you please create one?

File: src/app/components/file-uploader/file-uploader.component.ts [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+  templateUrl: './file-uploader.component.html',
+  styleUrls: ['./file-uploader.component.scss']
+})
+export class FileUploaderComponent {
+
+  @ViewChild('fileDropRef') fileInput: ElementRef;
+  @Input() acceptMultiple: boolean;
+  @Input() acceptedTypes: Array<String>;
+
+  private maxFileSize = 2097152;
```
Comment (Brend-Smits) [inline src/app/components/file-uploader/file-uploader.component.ts:26]: Comment above to describe what this means. Is this in bytes? Also might be nice to show this to the user.

File: src/app/components/file-uploader/file-uploader.component.ts [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  /**
+   * Converts the image to a base64 string
+   * @param file The file to convert
+   */
+  private generatePreview(file: UploadFile) {
+    // We can use a FileReader to generate a base64 string based on the image
+    const fileReader: FileReader = new FileReader();
+    // Convert the file to base64
```
Comment (Ruby77) [inline src/app/components/file-uploader/file-uploader.component.ts:116]: Sometimes you type comments this way:
`// comment`
But sometimes you do it this way:

```
/**
* Comment
*/
```
It's not really a problem, but I would suggest to use one style

(https://jsdoc.app/about-getting-started.html)

**Review by Ruby77 — COMMENTED** [review]
Looks awesome! However I have one small design question though:
image
image
Why would you use for uploading a circle shape, but in the real image a square shape?

File: src/app/components/file-uploader/file-uploader.component.ts (hunk 21 above)
Comment (waltersajtos) [reply src/app/components/file-uploader/file-uploader.component.ts:116]: I use // to comment in the middle of a function, it gets real messy and big if you use 3 lines for every comment. That's why I use both

**Review by Brend-Smits — APPROVED** [review]
Good work, thanks!
